### PR TITLE
[INFINITY-1449] CLI code refactoring

### PR DESCRIPTION
**Pull request by ssk2**
Merge after https://github.com/mesosphere/dcos-commons/pull/946 and https://github.com/mesosphere/dcos-commons/pull/951.

This PR refactors the flat directory structure of the CLI into three packages:
+ `commands`: containing handlers and methods for each subcommand.
+ `config`: containing a handful of constants and global variables that encapsulate configuration
+ `utils`: containing helper methods that facilitate HTTP communication, response parsing and integration with the DC/OS CLI.

I also updated the helloworld example to use this new layout. Will probably want to update this again the vendoring work is merged as well.

### CODE REVIEW

**Comment by benclarkwood** [comment]
@ssk2 Please resolve conflicts.

File: cli/commands/connection.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"fmt"
+	"github.com/mesosphere/dcos-commons/cli/utils"
+	"gopkg.in/alecthomas/kingpin.v2"
+	"strings"
+)
+
+// Connection section, manually implemented by some services (DEPRECATED, use common Endpoints)
+
+type ConnectionHandler struct {
```
Comment (benclarkwood) [inline cli/commands/connection.go]: Let's deprecate this now.

Comment (benclarkwood) [inline cli/commands/connection.go]: (everything is using HandleEndpoints AFAIK, and if they aren't: it's vendored.

File: cli/utils/http.go [hunk 2]
```diff
@@ -85,40 +67,40 @@ func HTTPPutJSON(urlPath, jsonPayload string) *http.Response {
 }
 
 func HTTPQuery(request *http.Request) *http.Response {
```
Comment (benclarkwood) [inline cli/utils/http.go]: Unless this is used outside of the package, HTTPQuery + CreateHTTPRequest and anything else only used in the package should be package private (lower case first letters).

Comment (benclarkwood) [inline cli/utils/http.go]: Looks like everything south of HTTPPutJSON should be private.

File: cli/utils/response.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package cli
+package utils
```
Comment (benclarkwood) [inline cli/utils/response.go]: `utils` is IMO the wrong name.

Everything in this package is for a client to the scheduler.

File: cli/config/config.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package config
```
Comment (benclarkwood) [inline cli/config/config.go]: Can config be made non-globals?

e.g. can it be created at the beginning and then passed around? (Makes DI easier in tests... ish)

File: cli/utils/response.go (hunk 3 above)
Comment (benclarkwood) [inline cli/utils/response.go]: Additionally, if the "client" can just be created once and re-used, then injected into commands (rather than globally referenced) writing tests will be much, much easier. (Not saying: make this change now, but it will eventually need to happen)

File: cli/config/config.go (hunk 4 above)
Comment (ssk2) [inline cli/config/config.go]: I think so, but this will require more refactoring. I'd like to do that in a separate PR if that is acceptable?

Comment (benclarkwood) [inline cli/config/config.go]: Yes, this can get bucketed into the same category as the comment above it (Will probably need to happen eventually but can be deferred for now).

File: cli/utils/response.go (hunk 3 above)
Comment (nickbp) [inline cli/utils/response.go]: As of this writing there are three files in utils/:
- dcoscli.go: access config data in the main CLI, run main CLI commands
- http.go: build requests to query the scheduler (and soon likely other services in the cluster)
- response.go: handle http responses

IMO two changes could be made here:
- response.go could be folded into http.go as they're pretty closely related (I think http.go even invokes response.go)
- I'm a bit hesitant to have a `utils` package in general. I think @benclarkwood 's point of instead calling it `client` or `clients` would work, as the functionality here is basically client code for querying 1) schedulers over HTTP and 2) the main DC/OS CLI over the commandline.

File: frameworks/helloworld/cli/dcos-hello-world/main.go [hunk 5]
```diff
@@ -2,17 +2,18 @@ package main
 
 import (
 	"github.com/mesosphere/dcos-commons/cli"
+    "github.com/mesosphere/dcos-commons/cli/commands"
```
Comment (nickbp) [inline frameworks/helloworld/cli/dcos-hello-world/main.go]: indent

File: tools/build_framework.sh [hunk 6]
```diff
@@ -79,7 +79,7 @@ case "$PUBLISH_STEP" in
         echo "---"
         echo "Build complete, skipping publish step."
         echo "Use one of the following additional arguments to get something that runs on a cluster:"
-        echo "- 'local': Host the build in a local HTTP server for use by a vagrant cluster."
+        echo "- 'local': Host the build in a local HTTP server for use by a DC/OS Vagrant cluster."
```
Comment (nickbp) [inline tools/build_framework.sh]: strictly speaking this was written for DC/OS Docker clusters but within a vagrant container. We found DC/OS Vagrant to not really be useful for our purposes (required huge machines when we're targeting e.g. laptops with 4-8GB)

see also: https://github.com/mesosphere/dcos-commons#quick-start

File: frameworks/helloworld/cli/dcos-hello-world/main.go [hunk 7]
```diff
@@ -2,17 +2,18 @@ package main
...
 
 func main() {
 	app := cli.New()
 
-	cli.HandleConfigSection(app)
-	cli.HandleEndpointsSection(app)
-	cli.HandlePlanSection(app)
-	cli.HandlePodsSection(app)
-	cli.HandleStateSection(app)
+	commands.HandleConfigSection(app)
```
Comment (nickbp) [inline frameworks/helloworld/cli/dcos-hello-world/main.go]: I assume other `main.go`s need the same update?

Comment (nickbp) [inline frameworks/helloworld/cli/dcos-hello-world/main.go]: Something to think about: These explicit section calls were added mainly to get around the downsides of unvendored dependencies. Now that vendoring is being introduced, we could bring back a single `HandleDefaultSections()` call that functions as a shortcut to all of these other calls.

File: cli/commands/connection.go (hunk 1 above)
Comment (ssk2) [inline cli/commands/connection.go]: Done.

File: cli/utils/http.go (hunk 2 above)
Comment (ssk2) [inline cli/utils/http.go]: Done.

File: cli/utils/response.go (hunk 3 above)
Comment (ssk2) [inline cli/utils/response.go]: I will rename this to `client` for now - we can fold response.go into http.go at a later point.

Comment (ssk2) [inline cli/utils/response.go]: Done.

File: frameworks/helloworld/cli/dcos-hello-world/main.go (hunk 7 above)
Comment (ssk2) [inline frameworks/helloworld/cli/dcos-hello-world/main.go]: Good point.

Comment (ssk2) [inline frameworks/helloworld/cli/dcos-hello-world/main.go]: I was only doing this for the example `helloworld` CLI - what's our policy about updating CLIs?

Comment (ssk2) [inline frameworks/helloworld/cli/dcos-hello-world/main.go]: Done (for helloworld).

File: tools/build_framework.sh (hunk 6 above)
Comment (ssk2) [inline tools/build_framework.sh]: Ah, okay. This might be confusing with both just `Vagrant` and `DC/OS Vagrant`. Would something like `local DC/OS cluster` work instead.

**Comment by ssk2** [comment]
@benclarkwood, @nickbp, rebased, conflicts resolved and most feedback addressed.

**Comment by ssk2** [comment]
retest this please

**Comment by benclarkwood** [comment]
Not sure what that merge issue is.

@jsrodman any thoughts?

**Comment by jsrodman** [comment]
This merge is odd.  Why is anything at the time of the go build trying to do a merge? is build.sh being invoked  recursively?

**Comment by ssk2** [comment]
retest this please

**Comment by ssk2** [comment]
retest this please
